File: src/main/java/org/javiermf/features/daos/ProductsDAO.java
package org.javiermf.features.daos;

import com.mysema.query.jpa.impl.JPADeleteClause;
import com.mysema.query.jpa.impl.JPAQuery;
import org.javiermf.features.exceptions.ObjectNotFoundException;
import org.javiermf.features.models.Feature;
import org.javiermf.features.models.Product;
import org.javiermf.features.models.QFeature;
import org.javiermf.features.models.QProduct;
import org.springframework.stereotype.Repository;
import org.springframework.transaction.annotation.Transactional;

import javax.persistence.EntityManager;
import javax.persistence.PersistenceContext;
import java.util.List;

@Repository
public class ProductsDAO {

    @PersistenceContext
    private EntityManager entityManager;

    QProduct qProduct = QProduct.product;
    QFeature qFeature = QFeature.feature;

    public List<Product> findAll() {
        JPAQuery query = new JPAQuery(entityManager);

        return query.from(qProduct).list(qProduct);
    }

    public Product findByName(String name) {
        JPAQuery query = new JPAQuery(entityManager);

        query.from(qProduct).where(qProduct.name.eq(name));
        Product product = query.singleResult(qProduct);
        if (product == null) {
            throw new ObjectNotFoundException(name);
        }
        return product;
    }

    @Transactional
    public void deleteByName(String productName) {
        Product product = findByName(productName);
        entityManager.remove(product);

    }

    @Transactional
    public void insert(Product product) {
        entityManager.persist(product);
    }

    @Transactional
    public void insertFeature(Feature feature) {
        entityManager.persist(feature);
    }

    @Transactional
    public void deleteFeature(Feature feature) {
        JPADeleteClause deleteClause = new JPADeleteClause(entityManager, qFeature);
        deleteClause.where(qFeature.id.eq(feature.getId()));
        deleteClause.execute();
    }
}

File: src/main/kotlin/org/javiermf/features/daos/ProductsDAO.kt
package org.javiermf.features.daos

import com.mysema.query.jpa.impl.JPADeleteClause
import com.mysema.query.jpa.impl.JPAQuery
import org.javiermf.features.exceptions.ObjectNotFoundException
import org.javiermf.features.models.Feature
import org.javiermf.features.models.Product
import org.javiermf.features.models.QFeature
import org.javiermf.features.models.QProduct
import org.javiermf.features.models.constraints.FeatureConstraint
import org.javiermf.features.models.constraints.QFeatureConstraint
import org.springframework.stereotype.Repository
import org.springframework.transaction.annotation.Transactional

import javax.persistence.EntityManager
import javax.persistence.PersistenceContext

@Repository
class ProductsDAO {

    @PersistenceContext
    private val entityManager: EntityManager? = null

    internal var qProduct = QProduct.product
    internal var qFeature = QFeature.feature
    internal var qFeatureConstraint = QFeatureConstraint.featureConstraint

    fun findAll(): List<Product> {
        val query = JPAQuery(entityManager!!)

        return query.from(qProduct).list(qProduct)
    }

    fun findByName(name: String): Product {
        val query = JPAQuery(entityManager!!)

        query.from(qProduct).where(qProduct.name.eq(name))
        return query.singleResult(qProduct) ?: throw ObjectNotFoundException(name)
    }

    @Transactional
    fun deleteByName(productName: String) {
        val product = findByName(productName)
        entityManager!!.remove(product)

    }

    @Transactional
    fun deleteConstraintsForProduct(productName: String) {
        val product = findByName(productName)
        for (featureConstraint in product.productFeatureConstraints) {
            entityManager!!.remove(featureConstraint)
        }
        product.productFeatureConstraints.clear()
    }

    @Transactional
    fun insert(product: Product) {
        entityManager!!.persist(product)
    }

    @Transactional
    fun insertFeature(feature: Feature) {
        entityManager!!.persist(feature)
    }

    @Transactional
    fun deleteFeature(feature: Feature) {
        val deleteClause = JPADeleteClause(entityManager!!, qFeature)
        deleteClause.where(qFeature.id.eq(feature.id!!))
        deleteClause.execute()
    }

    @Transactional
    fun insertConstraint(constraint: FeatureConstraint) {
        entityManager!!.persist(constraint)
    }

    @Transactional
    fun deleteConstraintForProduct(productName: String, constraintId: Long?) {
        val deleteClause = JPADeleteClause(entityManager!!, qFeatureConstraint)
        deleteClause.where(qFeatureConstraint.id.eq(constraintId!!))
        deleteClause.execute()
    }
}
